### PR TITLE
Added certificate refactor to gnoi_target

### LINKED ISSUE #153: gnoi_target if certificates are provided (ca, target & target_key) start in provisioned mode
Right now, gnoi_target is starting in bootstrapping mode by default.
However it should be possible to provide target certificates such that it starts in provisioned instead.

### PR DESCRIPTION
**Pull request by ericm**
Closes #153

### CODE REVIEW

**Comment by samribeiro** [comment]
1) Create a function like credentials.LoadCertificates() that returns nil, nil if no certificates were ever given instead of giving error.
2) Create a Settings struct for cert.Manager where you can pass the certificates and the bundle.
3) Note that certificates have the private key: https://godoc.org/crypto/tls#Certificate
4) In gnoi_target.start() before instantiating gNOIServer attempt loading the certificates with 1). Pass those certificates via Settings to gNOIServer.
5) at the end of gnoi_target.start() if the certificates were loaded call notifyCerts(1,1), else notifyCerts(0,0)

File: gnoi/cert/server.go [hunk 1]
```diff
@@ -336,3 +336,7 @@ func (s *Server) CanGenerateCSR(ctx context.Context, request *pb.CanGenerateCSRR
 	}
 	return ret, nil
 }
+
+func (s *Server) AddCertificates(certID string) (certs, caCerts int) {
```
Comment (samribeiro) [inline gnoi/cert/server.go:340]: Hum.. what does this do? =)

File: gnoi/gnoi.go [hunk 2]
```diff
@@ -45,30 +44,28 @@ type Server struct {
...
+		if err != nil {
+			return nil, fmt.Errorf("failed to generate private key: %v", err)
 		}
 		e, err := entity.CreateSelfSigned("gNOI server", privateKey)
 		if err != nil {
 			return nil, fmt.Errorf("failed to create self signed certificate: %v", err)
 		}
-		defaultCertificate = e.Certificate
+		certSettings.Cert = e.Certificate
+		certSettings.CertID = "0"
```
Comment (samribeiro) [inline gnoi/gnoi.go:58]: use the one in certSettings

File: gnoi_target/gnoi_target.go [hunk 3]
```diff
@@ -37,6 +40,7 @@ var (
 	muServe       sync.Mutex
 	bootstrapping bool
 
+	certID               = flag.String("cert_id", "", "Certificate ID for preloaded certificates")
```
Comment (samribeiro) [inline gnoi_target/gnoi_target.go:43]: use a default of "default"

File: gnoi_target/gnoi_target.go [hunk 4]
```diff
@@ -99,19 +105,30 @@ func start() {
...
-	bootstrapping = false
-	notifyCerts(0, 0) // Triggers bootstraping mode.
+	bootstrapping = certs != 0 && caCerts != 0
+	notifyCerts(certs, caCerts) // Triggers bootstraping mode.
 }
 
 // notifyReset is called when the factory reset service requires the server
 // to be restarted.
 func notifyReset() {
+	*certID = ""
```
Comment (samribeiro) [inline gnoi_target/gnoi_target.go:131]: no need, will use default above

File: utils/credentials/credentials.go [hunk 5]
```diff
@@ -97,8 +105,8 @@ func generateFromCA() ([]tls.Certificate, *x509.CertPool) {
 	return []tls.Certificate{*clientEnt.Certificate}, caPool
 }
 
-// LoadCertificates loads certificates from files or generates them from the CA.
-func LoadCertificates() ([]tls.Certificate, *x509.CertPool) {
+// ParseCertificates gets certificates from files or generates them from the CA.
+func ParseCertificates() ([]tls.Certificate, *x509.CertPool) {
```
Comment (samribeiro) [inline utils/credentials/credentials.go:109]: have this return:
([]*x509.Certificate, []*x509.Certificate)

File: utils/credentials/credentials.go [hunk 6]
```diff
@@ -107,10 +115,18 @@ func LoadCertificates() ([]tls.Certificate, *x509.CertPool) {
 			return generateFromCA()
 		}
 	}
-	log.Exit("Please provide -ca & -key or -ca, -cert & -ca_key")
 	return []tls.Certificate{}, &x509.CertPool{}
 }
 
+// LoadCertificates loads certificates from files and exits if there's an error.
+func LoadCertificates() ([]tls.Certificate, *x509.CertPool) {
+	certs, certPool := ParseCertificates()
```
Comment (samribeiro) [inline utils/credentials/credentials.go:123]: create the cert pool here

File: utils/credentials/credentials.go [hunk 7]
```diff
@@ -174,6 +193,11 @@ func ServerCredentials() []grpc.ServerOption {
 	}))}
 }
 
+// GetCACert returns the parsed CA cert.
```
Comment (samribeiro) [inline utils/credentials/credentials.go:196]: if you return the list of CA certificates above, then you dont need this

File: utils/credentials/credentials.go [hunk 8]
```diff
@@ -123,6 +139,8 @@ func ClientCredentials(server string) []grpc.DialOption {
 			tlsConfig.InsecureSkipVerify = true
 		} else {
 			certificates, certPool := LoadCertificates()
+			if len(certificates) == 0 {
```
Comment (samribeiro) [inline utils/credentials/credentials.go:142]: remove

File: utils/credentials/credentials.go [hunk 9]
```diff
@@ -148,6 +166,7 @@ func GetCAEntity() *entity.Entity {
 	if caEnt, err = entity.FromFile(*ca, *caKey); err != nil {
 		log.Exitf("Failed to load certificate and key from file: %v", err)
 	}
+	caCert = caEnt.Certificate.Leaf
```
Comment (samribeiro) [inline utils/credentials/credentials.go:169]: no need after the above change

File: gnoi/cert/manager.go [hunk 10]
```diff
@@ -40,6 +40,13 @@ type Info struct {
 // Notifier is called with number of Certificates and CA Certificates.
 type Notifier func(int, int)
 
+// ManagerSettings contains the certs and CA pool to be passed into the Manager.
+type ManagerSettings struct {
+	CertID   string
+	Cert     *tls.Certificate
```
Comment (samribeiro) [inline gnoi/cert/manager.go:46]: lets also use *x509.Certificate
see other comments

**Comment by ericm** [comment]
Everything's working now :)

File: gnoi/cert/manager.go [hunk 11]
```diff
@@ -52,11 +59,20 @@ type Manager struct {
 }
 
 // NewManager returns a Manager.
-func NewManager(privateKey crypto.PrivateKey) *Manager {
+func NewManager(settings *Settings) *Manager {
+	if settings == nil {
```
Comment (samribeiro) [inline gnoi/cert/manager.go:63]: No need to have this check, we should fail hard if this happens. If we want to have an empty Manager{} we can always create it without calling this function.

**Comment by ericm** [comment]
Removed :)

File: gnoi/gnoi.go [hunk 12]
```diff
@@ -45,30 +44,27 @@ type Server struct {
 }
 
 // NewServer returns a new server that can be used by the mock target.
-func NewServer(privateKey crypto.PrivateKey, defaultCertificate *tls.Certificate, resetSettings *reset.Settings, notifyReset reset.Notifier, osSettings *os.Settings) (*Server, error) {
-	if defaultCertificate == nil {
```
Comment (samribeiro) [inline gnoi/gnoi.go:49]: keep this logic within NewServer. Meaning, always create a private key and use it to generate a default certificate, then throw the private key away. This default certificate is only used to establish an encrypted session.
Pass the Settings untouched and directly to cert.NewManager.

File: utils/credentials/credentials.go [hunk 13]
```diff
@@ -67,38 +68,39 @@ func (a *userCredentials) RequireTransportSecurity() bool {
 }
 
 // loadFromFile loads the certificates from files.
```
Comment (samribeiro) [inline utils/credentials/credentials.go:70]: loads a certificate key pair into a tls certificate and a CA certificate into a x509 certificate.

File: utils/credentials/credentials.go [hunk 14]
```diff
@@ -67,38 +68,39 @@ func (a *userCredentials) RequireTransportSecurity() bool {
...
-func loadFromFile() ([]tls.Certificate, *x509.CertPool) {
+func loadFromFile() ([]tls.Certificate, []*x509.Certificate) {
```
Comment (samribeiro) [inline utils/credentials/credentials.go:71]: I think we can drop the slices, and jsut have pointers to tls and x509

File: utils/credentials/credentials.go [hunk 15]
```diff
@@ -67,38 +68,39 @@ func (a *userCredentials) RequireTransportSecurity() bool {
...
 	certificate, err := tls.LoadX509KeyPair(*cert, *key)
 	if err != nil {
 		log.Exitf("could not load client key pair: %s", err)
```
Comment (samribeiro) [inline utils/credentials/credentials.go:74]:
```
log.Exit("Could not load key/certificate pair from files:", err)
```

File: utils/credentials/credentials.go [hunk 16]
```diff
@@ -67,38 +68,39 @@ func (a *userCredentials) RequireTransportSecurity() bool {
...
 	}
-
-	certPool := x509.NewCertPool()
+	certificate.Leaf, err = x509.ParseCertificate(certificate.Certificate[0])
+	if err != nil {
+		log.Exitf("could not get leaf for certificate: %s", err)
```
Comment (samribeiro) [inline utils/credentials/credentials.go:78]:
```
log.Exit("Could not parse x509 certificate from tls certificate:", err)
```

File: utils/credentials/credentials.go [hunk 17]
```diff
@@ -67,38 +68,39 @@ func (a *userCredentials) RequireTransportSecurity() bool {
...
+	if err != nil {
+		log.Exit("Error parsing CA certificate", err)
 	}
-	return []tls.Certificate{certificate}, certPool
+	return []tls.Certificate{certificate}, []*x509.Certificate{caCert}
 }
 
 // generateFromCA generates a client certificate from the provided CA.
-func generateFromCA() ([]tls.Certificate, *x509.CertPool) {
+func generateFromCA() ([]tls.Certificate, []*x509.Certificate) {
```
Comment (samribeiro) [inline utils/credentials/credentials.go:93]: same, I think we can drop the slices

File: utils/credentials/credentials.go [hunk 18]
```diff
@@ -107,8 +109,18 @@ func LoadCertificates() ([]tls.Certificate, *x509.CertPool) {
 			return generateFromCA()
 		}
 	}
-	log.Exit("Please provide -ca & -key or -ca, -cert & -ca_key")
-	return []tls.Certificate{}, &x509.CertPool{}
+	return []tls.Certificate{}, []*x509.Certificate{}
```
Comment (samribeiro) [inline utils/credentials/credentials.go:112]: can also drop the slices

File: gnoi/cert/manager.go [hunk 19]
```diff
@@ -52,11 +59,17 @@ type Manager struct {
 }
 
 // NewManager returns a Manager.
-func NewManager(privateKey crypto.PrivateKey) *Manager {
+func NewManager(settings *Settings) *Manager {
+	certInfo := map[string]*Info{}
+	var privateKey crypto.PrivateKey
```
Comment (samribeiro) [inline gnoi/cert/manager.go:64]: Private key is always required. If tls.certificate is empty, then generate a key.

File: gnoi/gnoi.go (hunk 12 above)
Comment (ericm) [reply gnoi/gnoi.go:49]: I kept the logic the same from before. I'm still generating the private key and passing the resulting certificate to the settings.

File: utils/credentials/credentials.go [hunk 20]
```diff
@@ -43,6 +44,7 @@ var (
 	usernameKey    = "username"
 	passwordKey    = "password"
 	caEnt          *entity.Entity
+	targetName     = "client.com"
```
Comment (samribeiro) [inline utils/credentials/credentials.go:47]: we should make another PR turning this into an exportable flag that can be used by all the binaries

Comment (ericm) [reply utils/credentials/credentials.go:47]: Sure

**Comment by ericm** [comment]
Should be working now :)
